Category and subsystem is not being logged

pvollan opened the pull request:
#### 55d490a268c7a5fe34fc9ac10fcc54e8063e1e19
<pre>
Category and subsystem is not being logged
<a href="https://bugs.webkit.org/show_bug.cgi?id=253579">https://bugs.webkit.org/show_bug.cgi?id=253579</a>
rdar://106762697

Reviewed by Geoffrey Garen.

Category and subsystem is not being logged when the WebContent process is blocking access to logd in
the sandbox and forwarding logs to the Networking process. This patch sends the category and subsystem
along with the WebContent PID from the log hook to the Networking process, where the matching log
object will be used. A new function is added to do a sanity check of the strings sent. This patch does
not enable this feature, since we still need to evaluate if our logging needs are maintained.

* Source/WTF/wtf/spi/cocoa/OSLogSPI.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::validatedASCIIString):
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogHook):

Canonical link: <a href="https://commits.webkit.org/262295@main">https://commits.webkit.org/262295@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c92e2f019ad788b37d4882b8c711676edd2f0c4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1117 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1147 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1786 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/985 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1197 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1227 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1181 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1125 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1043 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1653 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1085 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1041 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1028 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/985 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1018 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1070 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2119 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1093 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1064 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/995 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1145 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1004 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1050 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/233 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/286 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1085 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1148 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/226 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1436,58 +1437,35 @@ void NetworkConnectionToWebProcess::installMockContentFilter(WebCore::MockConten
+void NetworkConnectionToWebProcess::logOnBehalfOfWebContent(const String& logChannel, const String& logCategory, const String& logString, uint8_t logType, int32_t pid)
+{
+    // Truncate long strings and only accept printable characters
+    CString string = validatedASCIIString(logString, 256);
```
What's the longest string we log today? Probably a limit like 2X that would be appropriate.

pvollan replied:
> What's the longest string we log today? Probably a limit like 2X that would be appropriate.

That is a good point! I have adjusted the limit accordingly.

Thanks for reviewing!

geoffreygaren commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -648,15 +648,20 @@ static void registerLogHook()
         return;
 
     os_log_set_hook(OS_LOG_TYPE_DEFAULT, ^(os_log_type_t type, os_log_message_t msg) {
+        if (msg->buffer_sz > 1024)
```
Do we need individual length limits if we have this overall limit too?

pvollan replied:
> Do we need individual length limits if we have this overall limit too?

I was thinking it could be good to have a length check in the Networking process too, in case the IPC endpoint is being accessed in other, unexpected ways.

Thanks for reviewing!

geoffreygaren reviewed (approved):
r=me

pvollan committed:
Category and subsystem is not being logged
https://bugs.webkit.org/show_bug.cgi?id=253579
rdar://106762697

Reviewed by Geoffrey Garen.

Category and subsystem is not being logged when the WebContent process is blocking access to logd in
the sandbox and forwarding logs to the Networking process. This patch sends the category and subsystem
along with the WebContent PID from the log hook to the Networking process, where the matching log
object will be used. A new function is added to do a sanity check of the strings sent. This patch does
not enable this feature, since we still need to evaluate if our logging needs are maintained.

* Source/WTF/wtf/spi/cocoa/OSLogSPI.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::validatedASCIIString):
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogHook):

Canonical link: https://commits.webkit.org/262295@main
